### PR TITLE
Introduce PolySharp

### LINKED ISSUE #3500: Investigate PolySharp package
https://github.com/Sergio0694/PolySharp

Rel : https://github.com/EasyNetQ/EasyNetQ.Management.Client/pull/213

As I see it allows to get rid of all our hacks with copying attributes.

### PR DESCRIPTION
**Pull request by sungam3r**
closes #3500

### CODE REVIEW

File: src/Directory.Build.props [hunk 1]
```diff
@@ -25,6 +25,10 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="PolySharp" Version="1.10.0" Condition="'$(TargetFramework)' != 'net5' AND '$(TargetFramework)' != 'net6' AND '$(TargetFramework)' != 'net7'">
```
Comment (Shane32) [inline src/Directory.Build.props:28]: Isn't the point here that PolySharp determines what polyfills are necessary based on which .NET version is being referenced?

```suggestion
    <PackageReference Include="PolySharp" Version="1.10.0">
```

Comment (Sergio0694) [reply src/Directory.Build.props:28]: It does indeed. This condition should not be necessary.

File: src/Directory.Build.props [hunk 2]
```diff
@@ -25,6 +25,10 @@
...
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
```
Comment (Sergio0694) [inline src/Directory.Build.props:30]: Can simplify this a bit:
```suggestion
      <IncludeAssets>build; analyzers</IncludeAssets>
```

Comment (sungam3r) [reply src/Directory.Build.props:30]: This is a default behavior when VS adds dev-only package. I would not change this.

File: src/Directory.Build.props (hunk 1 above)
Comment (sungam3r) [reply src/Directory.Build.props:28]: You may try. I tried it at first ang got compilation errors.

Comment (sungam3r) [reply src/Directory.Build.props:28]: OK, I got the reason.

**Comment by sungam3r** [comment]
@Shane32 I would not change in either way internal visibility for tests. Condition attribute looks like less evil.

**Comment by Sergio0694** [comment]
Note, if you also want attributes such as `[DynamicallyAccessedMembers]` to be generated, to help make the code simpler for #3461, you'll need to also define `PolySharpIncludeRuntimeSupportedAttributes` to `"true"` in a `PropertyGroup` 🙂

PolySharp only generates language support attributes by default, not the dummy runtime support ones too.

**Comment by Shane32** [comment]
Adding .NET 5.0 as an additional TFM for the MicrosoftDI and Dataloader projects solves the compilation issues so we can remove the conditions from the nuget package.  The issue has to do with the combination of having the InternalsVisibleTo attribute specified and the fact that some projects are targeting .NET Standard 2.0 while others are targeting .NET 5.0+ in the same build.

**Comment by Shane32** [comment]
> Note, if you also want attributes such as `[DynamicallyAccessedMembers]` to be generated, to help make the code simpler for #3461, you'll need to also define `PolySharpIncludeRuntimeSupportedAttributes` to `"true"` in a `PropertyGroup` 🙂
> 
> PolySharp only generates language support attributes by default, not the dummy runtime support ones too.

Thanks for the tip @Sergio0694 !

**Comment by Shane32** [comment]
@sungam3r Let's add this to `Tests.props` and remove the condition from the package reference:

```xml
<PackageReference Remove="PolySharp" />
```

**Comment by sungam3r** [comment]
https://github.com/graphql-dotnet/graphql-dotnet/pull/3502#issuecomment-1384376806

And that?

**Comment by Shane32** [comment]
Your call - I suggest fixing it either by (1) adding .NET 5 as a TFM for those projects, or (2) excluding PolySharp from Tests.props.  I suggest the latter for 3 reasons:

1. I don't think we need the polyfills for the tests
2. We would be required to have .NET 5 as a TFM for every project even if a .NET 5 TFM was unnecessary - for example if we drop specific support for .NET 5 (allowing .NET Standard 2.0 fallback)
3. There may be other issues when `PolySharpIncludeRuntimeSupportedAttributes` is enabled

**Review by Shane32 — APPROVED** [review]

**Comment by sungam3r** [comment]
I did not understand the issue with net5 magic fix but I would not add that TFM as well. I merge as is.

**Comment by sungam3r** [comment]
OK. It's time to tune #3641 👍

**Pull request merged**